lib/Sema/CSGen.cpp

```diff
@@ -356,7 +389,183 @@ namespace {
+        literalArgs.push_back(arg->getKind() >= ExprKind::First_LiteralExpr &&
+                              arg->getKind() <= ExprKind::Last_LiteralExpr);
+    } else {
+      literalArgs.push_back(
+          opExpr->getArg()->getKind() >= ExprKind::First_LiteralExpr &&
+          opExpr->getArg()->getKind() <= ExprKind::Last_LiteralExpr);
+    }
+    return literalArgs;
+  }
+
+  bool isTypeInCollectedTypes(ConstraintSystem &CS, LinkedTypeInfo &lti,
+                              TypeBase *type, bool literalArg) {
+    if (lti.collectedTypes.count(type))
+      return true;
+
+    if (auto generic = type->getAs<GenericTypeParamType>()) {
+      // If the generic param is an associated type or we have literals
+      // that may be lots of types, some of which may or may not
+      // conform, then this is too complicated, and assume it's
+      // possible.
+      if (literalArg || generic->getDepth() || generic->getIndex())
+        return true;
+
+      if (auto proto =
+              generic->getDecl()->getDeclContext()->getSelfProtocolDecl()) {
+        for (auto ty : lti.collectedTypes)
+          if (TypeChecker::conformsToProtocol(
+                  ty, proto, CS.DC, ConformanceCheckFlags::InExpression)
+                  .hasValue())
+            return true;
+      }
+      return false;
+    }
+
+    if (literalArg) {
+      for (auto proto : lti.literalProtocols(CS.TC.Context))
+        if (TypeChecker::conformsToProtocol(type, proto, CS.DC,
+                                            ConformanceCheckFlags::InExpression)
+                .hasValue())
+          return true;
+    }
+
+    for (auto convertTy : lti.collectedTypes) {
+      if (CS.TC.isConvertibleTo(convertTy, type, CS.DC))
+        return true;
+    }
+    return false;
+  }
+
+  bool expandCollectedTypesFromOperatorOverloads(ConstraintSystem &CS,
+                                                 LinkedTypeInfo &lti) {
+    while (1) {
+      llvm::SmallSet<TypeBase *, 4> additionalTypes;
+
+      for (auto opExpr : lti.operatorExprs) {
+        auto literalArgs = literalArgsForExpr(opExpr);
+        auto fn = opExpr->getFn();
+        ArrayRef<ValueDecl *> decls;
+
+        if (auto odre = dyn_cast<OverloadedDeclRefExpr>(fn))
+          decls = odre->getDecls();
+        else if (auto le = dyn_cast<LookupExpr>(fn))
+          decls = ArrayRef<ValueDecl *>(le->getMember().getDecl());
+        else
+          continue;
+
+        for (auto decl : decls) {
+          auto metaFnType = decl->getInterfaceType()->getAs<AnyFunctionType>();
+          auto fnType = metaFnType->getResult()->getAs<AnyFunctionType>();
+          if (!fnType)
+            fnType = metaFnType;
+          auto resultTy =
+              fnType->getResult()->getRValueType()->getDesugaredType();
+
+          if (lti.collectedTypes.count(resultTy) > 0)
+            continue;
+
+          bool possibleDecl = true;
+
+          if (auto generic = resultTy->getAs<GenericTypeParamType>()) {
```

Comment:
Just a note - overloads like `(Self, Self) -> Self` or `(T, T) -> T` are the biggest problem is at the moment - inability to filter out generic overloads which connect arguments to a result and prevent efficient solving of disconnected components, which means we have to explore a lot deeper into solution trees.

Reply:
This could be simplified down to 

```c++
if (llvm::any_of(fnType->getParams(), [&resultTy](const AnyFunctionType::Param &param) {
   return resultTy->isEqual(param.getPlainType());
})
  continue;
```

Reply:
> Just a note - overloads like `(Self, Self) -> Self` or `(T, T) -> T` are the biggest problem is at the moment - inability to filter out generic overloads which connect arguments to a result and prevent efficient solving of disconnected components, which means we have to explore a lot deeper into solution trees.

These types of overloads can't be ruled out as possibilities up front, nor disconnected, but this PR (by favoring choosing duplicate bindings of existing bindings and by merging those type variables once a solution exists) makes it so that the solution tree gets pruned to be much less branchy once we find a solution. The depth is unaffected, but the breadth gets vastly narrower.

Reply:
Sorry but I don't understand what you are trying to say, the problem is that in the worse case, especially when expression is invalid e.g `1 + "hello"`, we'd still get to explore almost everything just because generic overloads like `(Self, Self) -> Self` carry almost no context for the bindings and we can't verify conformances until generic parameter is actually bound to something.

Reply:
Well, with this PR, you'd disable every possible overload of `+` except for the generic `(Self, Self) -> Self` style, because all the non-generic ones are impossible. So you explore much less. I was trying to say that this is a big performance win, even when you can't do anything efficient with the generic style ones directly.

Reply:
What I'm trying to say is that presence of overloads with concrete types is not a problem if arguments are concrete as well or could be bound to something relatively easily.

Reply:
Like in the examples you have in the PR, all of the arguments are going to be concrete so it's pretty easy for the solver to avoid going deeper if argument conversions could be tested and rejected right away.